cpp/include/cudf/fixed_point/fixed_point.hpp

```diff
@@ -173,6 +224,153 @@ CUDF_HOST_DEVICE inline constexpr T shift(T const& val, scale_type const& scale)
   return left_shift<Rep, Rad>(val, scale);
 }
 
+/** @brief Extracts the sign, exponent, and integral significand of a floating point number.
+ *
+ * @note This returns all zero's for +/-0 and denormals, and zero's with an exponent of
+ * INT_MIN for the unrepresentable values +/-inf and NaN's.
+ *
+ * @tparam FloatingType Type of input floating point value
+ * @param floating The floating value to extract the components from
+ * @return A tuple containing the sign (bool), exponent (power-of-2), and integral significand.
+ */
+template <typename FloatingType,
+          typename cuda::std::enable_if_t<cuda::std::is_floating_point_v<FloatingType>>* = nullptr>
+CUDF_HOST_DEVICE inline auto extract_components(FloatingType floating)
+{
+  // This function is implemented using integer and bit operations, which is much
+  // faster than using the FPU functions for each component individually. This is especially
+  // true for doubles, as bottlenecking on the FP64 GPU pipeline can cut performance in half.
+
+  // See comments in FloatingComponentConstants about the floating point format.
+  using Constants    = FloatingComponentConstants<FloatingType>;
+  using IntegralType = typename Constants::IntegralType;
+
+  // Convert floating to integer
+  auto const integer_rep = [&]() {
+    IntegralType integer;
+    memcpy(&integer, &floating, sizeof(floating));
+    return integer;
+  }();
```

Comment:
I'll try this out and benchmark it.

Reply:
Actually, the order of bit fields (and how they are packed) is implementation-defined.  I'm not sure if it's a good idea to go down this road.  Bit fields are evil.

Reply:
If the order is implementation defined, indeed best not to do this. But I think there must be a clearer and more maintainable approach. I would suggest maybe making a class for the type conversion, with methods for retrieving the parts rather than just storing constants and using them in unabstracted computations.

Reply:
Retrieving the parts individually would incur multiple memcpy's.  We'll only need to be doing all of these at once.  And if you have a special value for the exponent (e.g. inf/nan/0/denormal) you don't want to query the other parts separately, you want to bail early, and the only way to encapsulate that knowledge is to have a extract-all function like this.

Reply:
OK. Assumming you have benchmarked the different ways to verify the compiler can't combine the memcpys.

Reply:
OK, so I did this test.  While testing I discovered that the early-return in extract_components() for 0/inf/NaN was actually a mistake due to the extra branching.  With that branch removed both this free-function method and using a class resulted in near-identical PTX code and benchmarks.  And the code itself is about the same number of lines.  I don't have a strong opinion on this, so I'll switch it over to using separate class methods.

Reply:
Sorry just see this. Why can't we just cast?
```
auto const integer_rep = IntegralType{*reinterpret_cast<IntegralType*>(&floating)};
```

Reply:
It's technically undefined behavior, and I think our compilation settings prohibit it.  In C++20 we can use std::bit_cast(), which is basically just a memcpy.

Reply:
Ah, strict aliasing rule, forgot about it. Thanks.